### PR TITLE
Energy compensation BSDFs refactor

### PR DESCRIPTION
**Pull request by est77**
- Introduced base classes for 2D and 3D albedo tables.
- Fixed issues with glossy and metal EC tables.
- Tables are not computed at startup anymore. They can be re-generated as part of unit tests.
- Fixed rare invalid samples when using the Beckmann NDF.
- Improved OSL glossy closure sampling weights using the albedo tables.
- Started Glass EC (work in progress).

### CODE REVIEW

File: src/appleseed/renderer/modeling/bsdf/energycompensation.cpp [hunk 1]
```diff
@@ -0,0 +1,485 @@
...
+            {
+                of << *p++ << "f";
+                if (i != last) of << ", ";
+
+                ++i;
+            }
+
+            of << endl;
+        }
+
```
Comment (dictoon) [inline src/appleseed/renderer/modeling/bsdf/energycompensation.cpp]: Remove one blank line.

File: src/appleseed/renderer/modeling/bsdf/energycompensation.cpp [hunk 2]
```diff
@@ -0,0 +1,485 @@
...
+
+        of << "};" << endl;
+    }
+}
+
+//
+// AlbedoTable3D class implementation.
```
Comment (dictoon) [inline src/appleseed/renderer/modeling/bsdf/energycompensation.cpp]: Fix wrong comment.

File: src/appleseed/renderer/modeling/bsdf/energycompensation.cpp [hunk 3]
```diff
@@ -0,0 +1,485 @@
...
+            }
+
+            of << endl;
+        }
+
+
+        of << "};" << endl;
+    }
+}
+
```
Comment (dictoon) [inline src/appleseed/renderer/modeling/bsdf/energycompensation.cpp]: Add one blank line before block comment.

File: src/appleseed/renderer/modeling/bsdf/energycompensation.cpp [hunk 4]
```diff
@@ -0,0 +1,485 @@
...
+    m_albedo_table = m_data;
+    m_avg_table = m_albedo_table + TableSize * TableSize;
+}
+
+AlbedoTable2D::AlbedoTable2D(const float* table)
+  : TableSize(32)
+  , TableHeight(TableSize + 1)
+  , m_data(nullptr)
+{
+    m_albedo_table = const_cast<float*>(table);
```
Comment (dictoon) [inline src/appleseed/renderer/modeling/bsdf/energycompensation.cpp]: That `const_cast<>` and the one below are suspicious. Why have a `const float*` parameter then?

File: src/appleseed/renderer/modeling/bsdf/glassbsdf.cpp [hunk 5]
```diff
@@ -834,7 +904,7 @@ namespace
             const float denom = abs(4.0f * wo.y * wi.y);
             if (denom == 0.0f)
             {
-                value.set(0.0f);
+                value = SpectrumType(0.0f);
```
Comment (dictoon) [inline src/appleseed/renderer/modeling/bsdf/glassbsdf.cpp]: That's expensive. Why the change?

File: src/appleseed/renderer/modeling/bsdf/glassbsdf.cpp [hunk 6]
```diff
@@ -261,6 +301,14 @@ namespace
                         wi,
                         sample.m_value.m_glossy,
                         sample.m_probability);
+
+                    add_energy_compensation_term(
+                        mdf,
+                        values,
+                        Vector3f(0.0f, 1.0f, 0.0f),
```
Comment (dictoon) [inline src/appleseed/renderer/modeling/bsdf/glassbsdf.cpp]: Why the vertical normal here? In what space is the normal expressed?

File: src/appleseed/renderer/modeling/bsdf/glassbsdf.cpp [hunk 7]
```diff
@@ -284,6 +332,14 @@ namespace
...
```
Comment (dictoon) [inline src/appleseed/renderer/modeling/bsdf/glassbsdf.cpp]: Same questions here.

File: src/appleseed/renderer/modeling/bsdf/glassbsdf.cpp [hunk 8]
```diff
@@ -932,9 +1002,385 @@ namespace
...
+        template <typename MDF>
+        static void add_energy_compensation_term(
+            const MDF&                  mdf,
+            const InputValues*          values,
+            const Vector3f&             normal,
+            const Vector3f&             outgoing,
+            const Vector3f&             incoming,
+            Spectrum&                   value)
+        {
+            if (values->m_energy_compensation != 0.0f)
```
Comment (dictoon) [inline src/appleseed/renderer/modeling/bsdf/glassbsdf.cpp]: Negate the test to remove one level of indentation.

File: src/appleseed/renderer/modeling/bsdf/glassbsdf.cpp [hunk 9]
```diff
@@ -932,9 +1002,385 @@ namespace
...
+            {
+                const float cos_in = dot(incoming, normal);
+                const float cos_on = dot(outgoing, normal);
+
+                const float roughness = values->m_roughness;
+
+                const float eta = values->m_precomputed.m_eta;
+                const float rcp_eta = 1.0f / values->m_precomputed.m_eta;
```
Comment (dictoon) [inline src/appleseed/renderer/modeling/bsdf/glassbsdf.cpp]: Are we not precomputing 1/eta already? If not, would it make sense?

File: src/appleseed/renderer/modeling/bsdf/glassbsdf.cpp [hunk 10]
```diff
@@ -932,9 +1002,385 @@ namespace
...
+                if (Eavg == 1.0f || Eavg_rcp_eta == 1.0f)
+                    return;
+
+                const float Favg = values->m_precomputed.m_fresnel_average;
+                const float Favg_rcp_eta = values->m_precomputed.m_fresnel_rcp_eta_average;
+
+                const bool is_entering = !values->m_precomputed.m_backfacing;
+                const bool is_reflection = cos_in * cos_on >= 0.0f;
+
+                Spectrum fms = is_reflection
```
Comment (dictoon) [inline src/appleseed/renderer/modeling/bsdf/glassbsdf.cpp]: We could save one spectrum copy by `madd()`'ing the right color at the end.

File: src/appleseed/renderer/modeling/bsdf/glassbsdf.cpp [hunk 11]
```diff
@@ -932,9 +1002,385 @@ namespace
...
+        {
+        }
+
+        template <typename MDF>
+        GlassAlbedoTable(const MDF& mdf, const float min_eta, const float max_eta)
+          : AlbedoTable3D(min_eta, max_eta)
+        {
+            for (size_t z = 0; z < TableSize; ++z)
+            {
+                const float eta = lerp(m_min_eta, m_max_eta, static_cast<float>(z) / (TableSize - 1));
```
Comment (dictoon) [inline src/appleseed/renderer/modeling/bsdf/glassbsdf.cpp]: Maybe `foundation::fit()` is a better... fit here?

File: src/appleseed/renderer/modeling/bsdf/glassbsdf.cpp [hunk 12]
```diff
@@ -932,9 +1002,385 @@ namespace
...
+      private:
+        // Compute the albedo for a given outgoing direction.
+        // See Physically Based Rendering, first edition, pp. 689-690.
+        template <typename MDF>
+        float compute_directional_albedo(
+            const float eta,
+            const float alpha,
+            const float cos_theta) const
+        {
+            // Special cases.
```
Comment (dictoon) [inline src/appleseed/renderer/modeling/bsdf/glassbsdf.cpp]: Singular?

File: src/appleseed/renderer/modeling/bsdf/glassbsdf.cpp [hunk 13]
```diff
@@ -932,9 +1002,385 @@ namespace
...
+                float probability = 0.0f;
+
+                GlassBSDFImpl::do_sample(
+                    mdf,
+                    s,
+                    false,
+                    shading_basis,
+                    alpha,
+                    alpha,
+                    1.0f,
```
Comment (dictoon) [inline src/appleseed/renderer/modeling/bsdf/glassbsdf.cpp]: I'd suggest adding comments next to these arguments to make it more readable.

File: src/appleseed/renderer/modeling/bsdf/glossybrdf.cpp [hunk 14]
```diff
@@ -368,10 +368,7 @@ namespace
                 }
                 break;
 
-              default:
-                assert(false);
-                pdf = 0.0f;
-                break;
+              assert_otherwise;
```
Comment (dictoon) [inline src/appleseed/renderer/modeling/bsdf/glossybrdf.cpp]: I'm not sure this change and other similar changes in this PR are a good idea, IIRC some compilers (gcc 7?) will complain that `pdf` is potentially uninitialized.

File: src/appleseed/renderer/modeling/bsdf/microfacethelper.cpp [hunk 15]
```diff
@@ -31,15 +31,16 @@
...
-#include "foundation/image/exrimagefilewriter.h"
-#include "foundation/image/image.h"
-#include "foundation/image/tile.h"
 #include "foundation/math/microfacet.h"
 #include "foundation/math/qmc.h"
 #include "foundation/math/sampling/mappings.h"
 #include "foundation/math/scalar.h"
 #include "foundation/math/vector.h"
 
+// appleseed.renderer headers.
```
Comment (dictoon) [inline src/appleseed/renderer/modeling/bsdf/microfacethelper.cpp]: Move before appleseed.foundation headers.

File: src/appleseed/renderer/modeling/bsdf/microfacethelper.cpp [hunk 16]
```diff
@@ -66,239 +67,128 @@ namespace
...
-            // Directional albedo.
-            const float* p = m_albedo_table;
-            for (size_t j = 0; j < TableSize; ++j)
-            {
-                for (size_t i = 0; i < TableSize; ++i)
-                {
-                    // Write 1 - E(u) to match the images in [2].
-                    image.set_pixel(i, j, Color3f(1.0f - p[i]));
-                }
+            if (alpha == 0.0f)
```
Comment (dictoon) [inline src/appleseed/renderer/modeling/bsdf/microfacethelper.cpp]: Fold into preceding test.

File: src/appleseed/renderer/modeling/bsdf/microfacethelper.h [hunk 17]
```diff
@@ -92,6 +92,10 @@ inline float highlight_falloff_to_gama(const float highlight_falloff)
     return foundation::mix(1.51f, 40.0f, t2 * t2 * t);
 }
 
+//
```
Comment (dictoon) [inline src/appleseed/renderer/modeling/bsdf/microfacethelper.h]: Add one additional blank line before block comment.

File: src/appleseed/renderer/modeling/bsdf/energycompensation.cpp (hunk 4 above)
Comment (est77) [reply src/appleseed/renderer/modeling/bsdf/energycompensation.cpp]: The tables can be computed or can be initialized with a C++ array now. We compute them in unit tests and use the C++ array version for normal usage. In one case, I need the table to be const and in the other I need to modify the table. That's the reason of the const_cast<>

File: src/appleseed/renderer/modeling/bsdf/glassbsdf.cpp (hunk 6 above)
Comment (est77) [reply src/appleseed/renderer/modeling/bsdf/glassbsdf.cpp]: wi and wo are both in local (shading basis) space.

File: src/appleseed/renderer/modeling/bsdf/glassbsdf.cpp (hunk 5 above)
Comment (est77) [reply src/appleseed/renderer/modeling/bsdf/glassbsdf.cpp]: evaluate_reflection and evaluate_refraction have to work with both Spectrums and floats.
I'll push something better.

File: src/appleseed/renderer/modeling/bsdf/glassbsdf.cpp (hunk 9 above)
Comment (est77) [reply src/appleseed/renderer/modeling/bsdf/glassbsdf.cpp]: We weren't precomputing it before. 
We are computing 1.0 / eta 2 times when energy compensation is on.

File: src/appleseed/renderer/modeling/bsdf/glassbsdf.cpp (hunk 11 above)
Comment (est77) [reply src/appleseed/renderer/modeling/bsdf/glassbsdf.cpp]: I find lerp easier to read than fit for this use case. And fix needs an extra static_cast<size_t>(0).

**Comment by est77** [comment]
Addressed review comments and rebased

**Comment by dictoon** [comment]
Thanks!
